src/internet/subsonic/subsonicdynamicplaylist.cpp

```diff
@@ -80,10 +81,87 @@ QNetworkReply* SubsonicDynamicPlaylist::Send(QNetworkAccessManager& network,
 }
 
 PlaylistItemList SubsonicDynamicPlaylist::Generate() {
-  return GenerateMore(kDefaultCount);
+  switch (type_) {
+    case QueryType_Album:
+      return GenerateMoreAlbums(kDefaultAlbumCount);
+    case QueryType_Song:
+      return GenerateMoreSongs(kDefaultSongCount);
+    default:
+      return GenerateMoreAlbums(kDefaultAlbumCount);
+  }
+}
+
+PlaylistItemList SubsonicDynamicPlaylist::GenerateMoreSongs(int count) {
+  SubsonicService* service = InternetModel::Service<SubsonicService>();
+  const int task_id =
+      service->app_->task_manager()->StartTask(tr("Fetching Playlist Items"));
+
+  BOOST_SCOPE_EXIT((service)(task_id)) {
+        // stop task when we're done
+        service->app_->task_manager()->SetTaskFinished(task_id);
+      }
+  BOOST_SCOPE_EXIT_END
+
+  QUrl url = service->BuildRequestUrl("getRandomSongs");
+  QNetworkAccessManager network;
+
+  if (count > kMaxCount) count = kMaxCount;
+
+  QUrlQuery url_query(url.query());
+  url_query.addQueryItem("size", QString::number(count));
+  url.setQuery(url_query);
+
+  PlaylistItemList items;
+
+  QNetworkReply* reply = Send(network, url, service->usesslv3_);
+  WaitForSignal(reply, SIGNAL(finished()));
+
+  reply->deleteLater();
+
+  if (reply->error() != QNetworkReply::NoError) {
+    qLog(Warning) << "HTTP error returned from Subsonic:"
+                  << reply->errorString() << ", url:" << url.toString();
+    return items;  // empty
+  }
+
+  QXmlStreamReader reader(reply);
+  reader.readNextStartElement();
+  if (reader.name() != "subsonic-response") {
+    qLog(Warning) << "Not a subsonic-response, aboring playlist fetch";
+    return items;
+  }
+
+  if (reader.attributes().value("status") != "ok") {
+    reader.readNextStartElement();
+    int error = reader.attributes().value("code").toString().toInt();
+    qLog(Warning) << "An error occurred fetching data.  Code: " << error
+                  << " Message: "
+                  << reader.attributes().value("message").toString();
```

Comment:
I think you want to `return` here.

Reply:
I copied that part from `GenerateMore` (which is now called `GenerateMoreAlbums`). I'll fix it in both.